rpm/cloud-hypervisor.spec

```diff
@@ -1,25 +1,32 @@
-# This spec file assumes you're building on an environment where:
-# * You have access to the internet during the build
-# * You have rustup installed on your system
-# * You have both x86_64-unknown-linux-gnu and  x86_64-unknown-linux-musl
-#   targets installed.
-
+# If this flag is set to 1, rustup installation needs to exist on the system and
+# <arch>-unknown-linux-gnu target is required.
+# If this flag is set to 0, distro specific Rust packages will be pulled into the build environment.
 %define using_rustup 1
+# If this flag is set to 1, <arch>-unknown-linux-musl target is required.
 %define using_musl_libc 1
+# If this flag is set to 1, the vendored crates archive and cargo.toml need to be prepared and
+# offline build is implied. Attached script update_src can be used for the vendorization.
+# If this flag is set to 0, access to the internet is required during the build.
+%define using_vendored_crates 0
```

Comment:
@sboeuf, @rbradford, I'd suggest having the vendored crates released as an additional release tarball, and then **alaways** relying on those here, avoiding the ifdef and having it as a second source.

Reply:
I'm no RPM expert at all, but the proposal sounds good to me.

@rbradford more of a general thought here but I think we should have @fidencio being pulled as a reviewer on any modification to the `rpm` folder, wdyt?

Reply:
> Would rustup play with that?

You have an excellent point here.  It wouldn't, and we'd need an online build for that.

Reply:
Looks like the situation is similar to the matrix below:

| tool | online | offline | vendor | 
|-------|--------|---------|--------|
| rustup | yes | no |  no
| rustc/cargo | yes | yes | yes |

An approach could be still to keep the current conditions for `using_vendored_crates` - if that's set, the user likely want's to use the vendor archive exclusively. Otherwise, one could remove `using_vendored_crates` as you've suggested, but introduce a separate one to explicitly tell whether build is offline.

There can be of course multiple solutions. The use case i was missing and that drew me is rustc/cargo+vendor+offline, that's why i've decided to condition this initially while keeping the original recipe unchanged. There can be other scenarios that users have, of course. In the end also - a suggestion to release the vendored crates IMO is great if doable, as it would keep CH releases buildable for some longer time disregarding the rapidly moving Rust environment.

Thanks

Reply:
Having `offline` build will be slightly more complicated as the distro must support installing the musl-libc in a distro-friendly way, which I am not sure if that's the case for CentOS / Fedora right now.

I'd suggest we go ahead with your approach for now, and then we work towards having the offline build figure out in a follow up PR.

Does this seem reasonable? @sboeuf @weltling

Reply:
Yes I agree. I mean if it's not straightforward to create this offline build approach, let's move on with this PR, and the improvements can definitely come later.

Reply:
> @sboeuf, @rbradford, I'd suggest having the vendored crates released as an additional release tarball, and then **alaways** relying on those here, avoiding the ifdef and having it as a second source.

You mean like this: https://github.com/cloud-hypervisor/cloud-hypervisor/releases/download/v23.1/cloud-hypervisor-v23.1.tar.xz ?

@bryteise already added it at while ago

Reply:
@fidencio thanks for mentioning `musl-libc`, it should be counted into the matrix. In general, it's not different as any other distro package dependency. Say if `rustc` is a dependency from the distro, its presence would be handled by the build tooling of the corresponding distro. Also, depending on the build automations, some could use SRPM cache or beyond mechanism in their pipelines. For `offline`, i'd rather refer to the Rust infrastructure itself, not the RPM build as a whole.

It looks like Fedora actually packages it https://src.fedoraproject.org/rpms/musl , but for any distro that uses a compatible package layout that should work a similar way. Perhaps we could say that taking a RHEL derivative as a package layout preference would be fine, things like build tooling and automation might be only taken into account occasionally and this should be distro agnostic in the spec. On my side, using an almost Fedora compatible distro now but it doesn't package `musl-libc` yet.

IMO going by the current approach should be fine for the time being. Also keeping the script might have its use. For example, one might still want to prepare their own vendored crates. Or, it is also possible that some would use a private crates registry, so then the sources might differ from the public ones. Please let me know otherwise, what would be seen as necessary to finalize this PR for the merge.

thanks!

Reply:
Cool, let's go ahead and merge it as it's and we can evaluate the next steps in a different issue.
Thanks @weltling!